File: packages/plot-routes/src/services/osrm.ts
import { Position } from 'Position'
import { Nearest } from 'OSRM'
import osrm from '../adapters/osrm'

const latLon = ({ lat, lon }: Position) => `${lon},${lat}`

export const nearest = async (position: Position) => {
  const { data }: { data: Nearest } = await osrm.get(`/nearest/v1/driving/${latLon(position)}`)

  return data
}

export default {
  nearest,
}

File: packages/engine_umbrella/apps/message_generator/lib/osrm.ex
defmodule MessageGenerator.Osrm do
  @osrmBase "https://osrm.iteamdev.io/"

  def nearest(%{lon: lon, lat: lat}) do
    HTTPoison.get!("#{@osrmBase}nearest/v1/driving/#{lon},#{lat}")
    |> Map.get(:body)
    |> Jason.decode!(keys: :atoms)
  end
end
